### ISSUE TITLE
Question about dataset size and model training 

### ISSUE THREAD

**Issue by hnsa9** [qcf-568/DocTamper#75]
Can I know if there's a recommended dataset size that is ideal for the model to converge? I have been training the model (both from scratch and using the pretrained doctamper model) using about 600+ ID image captured from mobile devices and then create the synthetic data using the same method as described in the paper but it seems that the model could not converge. 

I modified the dataset class to fit my custom dataset and wrapped the DTD model in a PyTorch Lightning module. 

This is my dataset class 
```

class ICTamperDataset(Dataset):
    def __init__(self, root_dir, mode, minq=95, train=True):
        self.root_dir = root_dir
        self.minq = minq
        self.mode = mode
        self.train = train
        
        self.image_list = sorted(os.listdir(os.path.join(root_dir, 'images')))
        self.label_list = sorted(os.listdir(os.path.join(root_dir, 'masks')))

        assert len(self.image_list) == len(self.label_list), "Number of images and labels must be the same"
        
        with open('qt_table.pk','rb') as fpk:
            pks = pickle.load(fpk)

        self.pks = {}
        for k,v in pks.items():
            self.pks[k] = torch.LongTensor(v)
     
        with open(f"{os.getcwd()}/pks/DocTamperV1-FCD_75.pk",'rb') as f: #  random selected compression factors /  keeping the random seed same all the test time
            self.record = pickle.load(f)

        if train:
            self.hflip = torchvision.transforms.RandomHorizontalFlip(p=1.0)
            self.vflip = torchvision.transforms.RandomVerticalFlip(p=1.0)
        
        self.totsr = ToTensorV2()
        self.toctsr = torchvision.transforms.Compose([torchvision.transforms.ToTensor(),torchvision.transforms.Normalize(mean=(0.485, 0.455, 0.406), std=(0.229, 0.224, 0.225))])


    def __len__(self):
        return len(self.image_list)

    def __getitem__(self, index):
        # Load image
        img_path = os.path.join(self.root_dir, 'images', self.image_list[index])
        im = Image.open(img_path) #.convert('RGB')
        
        lbl_path = os.path.join(self.root_dir, 'masks', self.label_list[index])
        
        with open(lbl_path, 'rb') as file:
            lbl_buf = file.read()

        mask = (cv2.imdecode(np.frombuffer(lbl_buf,dtype=np.uint8),0)!=0).astype(np.uint8) # mask image 
        
        H,W = mask.shape
        record = self.record[index]
        choicei = len(record)-1
        q = int(record[-1])
        use_qtb = self.pks[q]
        if choicei>1:
            q2 = int(record[-3])
            use_qtb2 = self.pks[q2]
        if choicei>0:
            q1 = int(record[-2])
            use_qtb1 = self.pks[q1]
        mask = self.totsr(image=mask.copy())['image']
        with tempfile.NamedTemporaryFile(delete=True) as tmp:
            im = im.convert("L")
            if choicei>1:
                im.save(tmp,"JPEG",quality=q2)
                im = Image.open(tmp)
            if choicei>0:
                im.save(tmp,"JPEG",quality=q1)
                im = Image.open(tmp)
            im.save(tmp,"JPEG",quality=q)
            jpg = jpegio.read(tmp.name)
            dct = jpg.coef_arrays[0].copy()
            im = im.convert('RGB')        
            
        return {
            'image': self.toctsr(im),
            'label': mask.long(),
            'rgb': np.clip(np.abs(dct),0,20),
            'q':use_qtb,
            'i':q
        }
```

This is my PyTorch Lightning DTD model class
```

class seg_dtd_lightning(pl.LightningModule):
    def __init__(self, model_name='resnet18', n_class=1, learning_rate=1e-3, gpu=True):
        super().__init__()
        self.gpu = gpu
        self.model = DTD(encoder_name=model_name, classes=n_class)
        
        if n_class == 1:
            mode ='binary'
        else:
            mode = 'multiclass'
   
        self.criterion = LovaszLoss(mode=mode)
        self.learning_rate = learning_rate
  

    def load_model(self, path):
        location = 'cuda' if self.gpu else 'cpu'
        self.model.load_state_dict(torch.load(path, map_location=location), strict=False)

    @autocast()
    def forward(self, x, dct, qt):
        x = self.model(x, dct, qt)
        return x

    def training_step(self, batch, batch_idx):
        device = 'cuda'
        data, target, dct_coef, qs, q = batch['image'], batch['label'],batch['rgb'], batch['q'],batch['i']
        data, target, dct_coef, qs = Variable(data.to(device)), Variable(target.to(device)), Variable(dct_coef.to(device)), Variable(qs.unsqueeze(1).to(device))
        y_hat = self(data,dct_coef,qs)
        loss = self.criterion(y_hat, target)
        return loss

    def validation_step(self, batch, batch_idx):
        device = 'cuda'
        data, target, dct_coef, qs, q = batch['image'], batch['label'],batch['rgb'], batch['q'],batch['i']
        data, target, dct_coef, qs = Variable(data.to(device)), Variable(target.to(device)), Variable(dct_coef.to(device)), Variable(qs.unsqueeze(1).to(device))
        y_hat = self(data,dct_coef,qs)
        loss = self.criterion(y_hat, target)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.learning_rate)
        return optimizer

```

My training code
```

    dataset_folder = f"{os.getcwd()}/data/DocTamper"
    ic_dataset = ICTamperDataset(dataset_folder,False,minq=args.minq)

    train_size = int(0.8* len(ic_dataset))
    test_size = len(ic_dataset) - train_size

    print("Train dataset size:", train_size)
    print("Test dataset size:", test_size)

    train_dataset, test_dataset = torch.utils.data.random_split(ic_dataset, [train_size, test_size])

    batch_size = 6
    workers = 2
    train_loader = DataLoader(dataset=train_dataset, batch_size=batch_size, num_workers=workers, shuffle=True)  
    test_loader = DataLoader(dataset=test_dataset, batch_size=batch_size, num_workers=workers, shuffle=False) 

    pth_file = args.pth
    train = True

    use_pretrained = False

    # Train using pytorch lightning model class
    model = seg_dtd_lightning('', n_class=2).to('cuda') 

    if use_pretrained:
        model.load_model(pth_file) 

    trainer = pl.Trainer(max_epochs=50, gpus=[0], precision=16, strategy='dp') 
    trainer.fit(model, train_loader, test_loader)
    torch.save({'state_dict':trainer.model.state_dict()}, "pths/doctamper_pl_2.pth")

```

If you have any input or advice regarding this, it would be appreciated. I'm not sure if it's due to my dataset or the way that I train it. 

Thank you in advance!

**Comment by qcf-568** [comment]
Hello, thanks for your attention in our work.
Can you reproduce the results on DocTamper and T-SROIE datasets with this modified code?
How many iterations have your trained the model?
How to judge that model does not converge?
What is the original format of the images before tampering?
What is the pre-process for the image before tampering?
What is the pre-process of the input to fit the fixed size (e.g., 512 * 512) during training and inference? Crop or resize?

**Comment by hnsa9** [comment]
**Can you reproduce the results on DocTamper and T-SROIE datasets with this modified code?**
I have tried with DocTamperV1-FCD data but was not able to reproduce the result 

**How many iterations have your trained the model?**
I have trained it on 30 epochs 

**How to judge that model does not converge?**
I evaluated the model with the existing evaluation code, which produces the iou, precision and recall. IoU is 0.007, precision is 0.03542 and recall is 0.999. The train and validation loss is also stagnant at 0.500~

**What is the original format of the images before tampering?**
The original format of the images are jpeg 

**What is the pre-process for the image before tampering?**
No preprocess is done before tampering except cropping the image into patches using this script https://github.com/qcf-568/DocTamper/blob/main/models/tsroie/infer_sroie.py. After that tampering is done on the patches and patches are used to train the model

**What is the pre-process of the input to fit the fixed size (e.g., 512 * 512) during training and inference? Crop or resize?**
No preprocess

I hope it answers your questions

Additionally,  I tried to create a normal training loop without wrapping the model in lightning and it seems to work? At least the loss value is not stagnant but I was not able to calculate the loss using LovaszLoss, it keeps throwing error like size mismatch of tensor a and tensor b so I just used soft cross entropy loss

Just to confirm, the loss value during training is lovasz loss + soft cross entropy loss right?

**Comment by qcf-568** [comment]
If you cannot reproduce the results, you can email to me with your edu email address to request the training pipeline.

30 epochs are too short for a 600+ image dataset to converge. T-SROIE dataset has 600+ images, with a high JPEG quality of 95, I even train the model for 400~800 epochs.

Lovasz loss is not necessary, I used it just as previous open-source projects from others. If a model dose not converge, it's not the problem of pure cross entropy loss.

**Comment by hnsa9** [comment]
thank you for your response, i will email you for the training pipeline and take a look at it
